client/src/pages/SignIn/SignIn.tsx

```diff
@@ -2,13 +2,31 @@ import React from 'react';
 import styles from './SignIn.module.css';
 import logo from './TUTVLogo.png';
 import Button from 'components/Button';
+import { useStore } from '../../store';
+import { useApiRequest } from '../../api';
 
 const SignIn: React.FC = () => {
   const [username, setUsername] = React.useState('');
   const [password, setPassword] = React.useState('');
+  const { dispatch } = useStore();
+
+  const getToken = useApiRequest('token', {
+    method: 'POST',
+    body: JSON.stringify({ username, password }),
+    headers: { 'Content-Type': 'application/json' },
+  });
+
+  const doSignIn = (tokens: { refresh: string; access: string }) => {
+    const { refresh, access } = tokens;
+    dispatch({ type: 'login', tokens: { refresh, access } });
+  };
 
   const attemptSignIn = () => {
-    console.log('Username:', username, 'Password:', password);
+    getToken()
+      .then(doSignIn)
+      .catch((e) => {
+        alert(`Something went wrong: ${e.response && e.response.detail}`);
+      });
```

Comment:
`useApiRequest` doesn't return a promise, it returns a function. That function, when called, performs a network request and returns a promise. To my knowledge, hooks that return functions are not an anti-pattern; the `useState` hook returns a setter function, for example.

It's important to have this mechanism as a part of our API because it enables **imperative** API access. The `swr` model is very declarative: you outline what data the component needs, and the requests are made automatically; the declarative model abstracts away the requests themselves and just returns the resulting data.

However, we do need to make some requests (like this login request) imperatively, in order to be able to explicitly control when they are made. For example, in this case, the request is made in response to a button press. This is impossible in a declarative pattern; we need to be able to say imperatively "ok, make the request *now*" when the button is pressed, and not before or after.

Reply:
https://github.com/zeit/swr/issues/10

Reply:
that still poses fetching as a fundamentally declarative problem, and delays the request by waiting until we "declare" the request to be ready. I think that in cases when we want to make an imperative data request, we should just do it imperatively, instead of going out of our way to find an API to do it with a declarative mechanism that doesn't really "fit."